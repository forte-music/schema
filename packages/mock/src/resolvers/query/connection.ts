import { Connection } from '../../models';

export interface ConnectionArgs {
  after?: string;
  first?: number;
}

export const handleConnection = <InputType, NodeType>(
  keys: InputType[],
  getNode: (key: InputType) => NodeType,
  { first = 25, after }: ConnectionArgs
): Connection<NodeType> => {
  const lowerBound = after ? parseInt(after, 10) + 1 : 0;
  const upperBound = first === -1 ? keys.length : lowerBound + first;

  const acceptedKeys = keys.slice(lowerBound, upperBound);

  return {
    count: keys.length,
    pageInfo: {
      hasNextPage: upperBound < keys.length,
    },
    edges: acceptedKeys.map((key, index) => ({
      cursor: (index + lowerBound).toString(),
      node: getNode(key),
    })),
  };
};
