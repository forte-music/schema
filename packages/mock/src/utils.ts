type Identifiable = {
  id: string;
};

export const mustGet = <K, V>(map: Map<K, V>, key: K): V => {
  const value = map.get(key);

  if (value === undefined) {
    throw new TypeError(`Schema invalid. Value not found in map. Key: ${key}`);
  }

  return value;
};

export const mustGetKeys = <T>(map: Map<string, T>, keys: string[]): T[] =>
  keys.map(key => mustGet(map, key));

// The getMap is needed because due to circular imports, the map may be
// undefined at import time. For example if song imports album's map and
// album import's songs map album's song map will be undefined because song
// hasn't exported anything yet.
export const arrayPropertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  keys: string[]
) => ({
  get: () => mustGetKeys(getMap(), keys),
});

export const propertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
});

export const propertyDescriptorWithSet = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
  set: (val: T): void => {
    getMap().set(key, val);
  },
});

export const makeMap = <T extends Identifiable>(list: T[]): Map<string, T> =>
  list.reduce((map, identifiable: T) => {
    addToMap(map, identifiable);
    return map;
  }, new Map());

export const addToMap = <T extends Identifiable>(
  map: Map<string, T>,
  item: T
) => {
  map.set(item.id, item);
};

export const now = () => Math.floor(Date.now() / 1000);

export const uuidForNum = (num: number): string =>
  num
    .toString(16)
    .toLowerCase()
    .padStart(32, '0');
