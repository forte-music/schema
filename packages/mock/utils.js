// @flow
type Identifiable = {
  id: string,
};

export const mustGet = <K: string, V>(map: Map<K, V>, key: K): V => {
  const value = map.get(key);

  if (value === undefined) {
    throw new TypeError(`Schema invalid. Value not found in map. Key: ${key}`);
  }

  return value;
};

export const mustGetKeys = <T>(map: Map<string, T>, keys: string[]): T[] =>
  keys.map(key => mustGet(map, key));

// The getMap is needed because due to circular imports, the map may be
// undefined at import time. For example if song imports album's map and
// album import's songs map album's song map will be undefined because song
// hasn't exported anything yet.
export const arrayPropertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  keys: string[]
) => ({
  get: () => mustGetKeys(getMap(), keys),
});

export const propertyDescriptor = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
});

export const propertyDescriptorWithSet = <T>(
  getMap: () => Map<string, T>,
  key: string
) => ({
  get: () => mustGet(getMap(), key),
  set: (val: T): void => {
    getMap().set(key, val);
  },
});

export const makeMap = <T: Identifiable>(list: T[]): Map<string, T> =>
  list.reduce((map, identifiable: T) => {
    map.set(identifiable.id, identifiable);
    return map;
  }, new Map());
