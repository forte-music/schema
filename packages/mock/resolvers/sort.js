// @flow
import type { Connection, UserStats } from '../models';
import type { ConnectionArgs } from './connection';
import { handleConnection } from './connection';

type SortBy =
  | 'RECENTLY_ADDED'
  | 'LEXICOGRAPHICALLY'
  | 'RECENTLY_PLAYED'
  | 'RELEVANCE';

type SortParams = {
  sortBy?: SortBy,
  reverse?: boolean,
  filter?: string,
};

type Sortable = {
  name: string,
  stats: UserStats,
  timeAdded: number,
};

type LinkedSortInfo<T> = {
  sortable: Sortable,
  original: T,
};

const compare = (a: any, b: any): number => {
  if (a > b) {
    return 1;
  }

  if (a === b) {
    return 0;
  }

  if (b > a) {
    return -1;
  }

  throw new TypeError(`Invariant a: ${a}, b: ${b}`);
};

// Returns how closely the record matches the query string.
const searchDistance = (record: string, query: string): number => {
  const normalizedQuery = query.toLowerCase().replace(/ /g, '');
  const normalizedRecord = record.toLowerCase();

  let strength = 0;
  let currentStrength = 0;
  let lastMatchIndex = 0;

  const khars = normalizedQuery.split('');
  for (const khar of khars) {
    const indexOfQueryCharacter = normalizedRecord.indexOf(
      khar,
      lastMatchIndex
    );
    if (indexOfQueryCharacter === -1) {
      // A character from the query isn't in the record. Sort to bottom.
      return -1;
    }

    if (lastMatchIndex === indexOfQueryCharacter - 1) {
      // continuation of last match
      currentStrength++;
    } else {
      // broken streak, reset streak strength
      currentStrength = 0;
    }

    // move last match index
    lastMatchIndex = indexOfQueryCharacter;

    // update strength
    strength += currentStrength;
  }

  return strength;
};

const reverseCompare = (a: any, b: any): number => reverse(a, b, compare);

const reverse = <T, R>(a: T, b: T, inner: (a: T, b: T) => R): R => inner(b, a);

const identity = <T, R>(a: T, b: T, inner: (a: T, b: T) => R): R => inner(a, b);

type Comparator<T> = (T, T) => number;

const sortWithReverse = <T>(
  reversed: boolean,
  inner: Comparator<T>
): Comparator<T> => {
  return (a: T, b: T): number => {
    const transform = reversed ? reverse : identity;
    return transform(a, b, inner);
  };
};

export type SortConnectionArgs = ConnectionArgs & {
  sort?: SortParams,
};

export const listItemsResolver = <T>(
  values: T[],
  { sort, ...connectionArgs }: SortConnectionArgs,
  process: (item: T) => Sortable
): Connection<T> => {
  const localHandleConnection = (items: T[]) =>
    handleConnection(items, value => value, connectionArgs);

  if (!sort) {
    // No sort was specified, use default sort.

    return localHandleConnection(values);
  }

  const { sortBy = 'LEXICOGRAPHICALLY', reverse = false, filter = '' } = sort;

  let sorter: Comparator<Sortable> = (a: Sortable, b: Sortable): number => 0;
  let filterFunc = (item: Sortable) => searchDistance(item.name, filter) !== -1;

  switch (sortBy) {
    case 'RECENTLY_ADDED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.timeAdded, b.timeAdded);
      break;

    case 'LEXICOGRAPHICALLY':
      sorter = (a: Sortable, b: Sortable) => compare(a.name, b.name);
      break;

    case 'RECENTLY_PLAYED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.stats.lastPlayed || 0, b.stats.lastPlayed || 0);
      break;

    case 'RELEVANCE':
      if (!filter) {
        throw new TypeError('Used relevance sort without filter.');
      }

      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(
          searchDistance(a.name, filter),
          searchDistance(b.name, filter)
        );

      break;

    default:
      throw new TypeError(`Unknown Sort ${sortBy}`);
  }

  const mappedValues: LinkedSortInfo<T>[] = values.map(item => ({
    original: item,
    sortable: process(item),
  }));

  const filteredValues = mappedValues.filter(({ original: T, sortable }) =>
    filterFunc(sortable)
  );

  const sortedValues = filteredValues.slice();
  sortedValues.sort(
    sortWithReverse(
      reverse,
      (
        { sortable: a }: LinkedSortInfo<T>,
        { sortable: b }: LinkedSortInfo<T>
      ) => sorter(a, b)
    )
  );

  return localHandleConnection(sortedValues.map(({ original }) => original));
};

export const itemsResolver = <T: Sortable>(map: Map<string, T>) => (
  _: void,
  args: SortConnectionArgs
): Connection<T> => {
  const values = Array.from(map.values());
  return listItemsResolver(values, args, item => item);
};
