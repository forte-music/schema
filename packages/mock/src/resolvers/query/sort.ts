import { Connection } from '../../models';
import { ConnectionArgs, handleConnection } from './connection';

type SortBy = 'RECENTLY_ADDED' | 'LEXICOGRAPHICALLY' | 'RECENTLY_PLAYED';

interface SortParams {
  sortBy?: SortBy;
  reverse?: boolean;
  filter?: string;
}

interface Sortable {
  name: string;
  lastPlayed?: number;
  timeAdded: number;
}

interface LinkedSortInfo<T> {
  sortable: Sortable;
  original: T;
}

interface Comparator<T> {
  (a: T, b: T): number;
}

export const compare = <T>(a: T, b: T): number => {
  if (a > b) {
    return 1;
  }

  if (a === b) {
    return 0;
  }

  if (b > a) {
    return -1;
  }

  throw new TypeError(`Invariant a: ${a}, b: ${b}`);
};

export const reverseCompare = <T>(a: T, b: T): number => reverse(a, b, compare);

const reverse = <T>(a: T, b: T, inner: Comparator<T>): number => inner(b, a);

const identity = <T>(a: T, b: T, inner: Comparator<T>): number => inner(a, b);

// Returns how closely the record matches the query string.
const searchDistance = (record: string, query: string): number => {
  const normalizedQuery = query.toLowerCase().replace(/ /g, '');
  const normalizedRecord = record.toLowerCase();

  let strength = 0;
  let currentStrength = 0;
  let lastMatchIndex = 0;

  const khars = normalizedQuery.split('');
  for (const khar of khars) {
    const indexOfQueryCharacter = normalizedRecord.indexOf(
      khar,
      lastMatchIndex
    );
    if (indexOfQueryCharacter === -1) {
      // A character from the query isn't in the record. Sort to bottom.
      return -1;
    }

    if (lastMatchIndex === indexOfQueryCharacter - 1) {
      // continuation of last match
      currentStrength++;
    } else {
      // broken streak, reset streak strength
      currentStrength = 0;
    }

    // move last match index
    lastMatchIndex = indexOfQueryCharacter;

    // update strength
    strength += currentStrength;
  }

  return strength;
};

const sortWithReverse = <T>(
  reversed: boolean,
  inner: Comparator<T>
): Comparator<T> => {
  return (a: T, b: T): number => {
    const transform = reversed ? reverse : identity;
    return transform(a, b, inner);
  };
};

export type SortConnectionArgs = ConnectionArgs & {
  sort?: SortParams;
};

export const listItemsResolver = <T>(
  values: T[],
  { sort, ...connectionArgs }: SortConnectionArgs,
  process: (item: T) => Sortable
): Connection<T> => {
  const localHandleConnection = (items: T[]) =>
    handleConnection(items, value => value, connectionArgs);

  if (!sort) {
    // No sort was specified, use default sort.

    return localHandleConnection(values);
  }

  const { sortBy = 'LEXICOGRAPHICALLY', reverse = false, filter = '' } = sort;

  let sorter: Comparator<Sortable> = () => 0;
  let filterFunc = (item: Sortable) => searchDistance(item.name, filter) !== -1;

  switch (sortBy) {
    case 'RECENTLY_ADDED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.timeAdded, b.timeAdded);
      break;

    case 'LEXICOGRAPHICALLY':
      sorter = (a: Sortable, b: Sortable) => compare(a.name, b.name);
      break;

    case 'RECENTLY_PLAYED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.lastPlayed || 0, b.lastPlayed || 0);
      break;

    default:
      throw new TypeError(`Unknown Sort ${sortBy}`);
  }

  const mappedValues: LinkedSortInfo<T>[] = values.map(item => ({
    original: item,
    sortable: process(item),
  }));

  const filteredValues = mappedValues.filter(({ sortable }) =>
    filterFunc(sortable)
  );

  const sortedValues = filteredValues.slice();
  sortedValues.sort(
    sortWithReverse(
      reverse,
      (
        { sortable: a }: LinkedSortInfo<T>,
        { sortable: b }: LinkedSortInfo<T>
      ) => sorter(a, b)
    )
  );

  return localHandleConnection(sortedValues.map(({ original }) => original));
};

export const itemsResolver = <T extends Sortable>(map: Map<string, T>) => (
  _: void,
  args: SortConnectionArgs
): Connection<T> => {
  const values = Array.from(map.values());
  return listItemsResolver(values, args, item => item);
};
