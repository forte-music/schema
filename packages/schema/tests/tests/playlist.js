import gql from 'graphql-tag';
import { print } from 'graphql/language/printer';

import { testConnection } from '../connection';
import { testSort } from '../sort';

import PlaylistFields from './fragments/PlaylistFields.graphql';
import PlaylistItemFields from './fragments/PlaylistItemFields.graphql';
import SongFields from './fragments/SongFields.graphql';
import UserStatsFields from './fragments/UserStatsFields.graphql';

import client from '../client';
import { uuidForNum } from '../utils';

it('should get a playlist by id', async () => {
  const variables = { playlistId: uuidForNum(1) };

  const query = gql`
    query($playlistId: ID!) {
      playlist(id: $playlistId) {
        ...PlaylistFields

        items(first: 10) {
          edges {
            node {
              ...PlaylistItemFields

              song {
                ...SongFields
              }
            }
          }
        }
      }
    }

    ${PlaylistFields}
    ${PlaylistItemFields}
    ${SongFields}
  `;

  const resp = await client.request(print(query), variables);
  expect(resp).toMatchSnapshot();
});

const PlaylistFragment = gql`
  fragment PlaylistFragment on Playlist {
    ...PlaylistFields

    items(first: -1) {
      edges {
        node {
          song {
            id
          }
        }
      }
    }
  }

  ${PlaylistFields}
`;

const expectMatchesPlaylist = async (playlistId, expectedPlaylist) => {
  const query = gql`
    query($playlistId: ID!) {
      playlist(id: $playlistId) {
        ...PlaylistFragment
      }
    }

    ${PlaylistFragment}
  `;

  const { playlist: queriedPlaylist } = await client.request(print(query), {
    playlistId,
  });

  expect(queriedPlaylist).toMatchObject({
    id: playlistId,
    ...expectedPlaylist,
  });
};

testConnection('playlist', async ({ first, after }) => {
  const query = gql`
    query($playlistId: ID!, $first: Int, $after: String) {
      playlist(id: $playlistId) {
        items(first: $first, after: $after) {
          count
          pageInfo {
            hasNextPage
          }

          edges {
            cursor

            node {
              id
            }
          }
        }
      }
    }
  `;

  const { playlist: { items } } = await client.request(print(query), {
    playlistId: uuidForNum(3),
    first,
    after,
  });

  return items;
});

testConnection('playlists', async ({ first, after }) => {
  const query = gql`
    query($first: Int, $after: String) {
      playlists(first: $first, after: $after) {
        count
        pageInfo {
          hasNextPage
        }

        edges {
          cursor

          node {
            id
          }
        }
      }
    }
  `;

  const { playlists } = await client.request(print(query), {
    first,
    after,
  });

  return playlists;
});

testSort('playlists', async ({ sortBy, reverse }) => {
  const query = gql`
    query($sortBy: SortBy!, $reverse: Boolean!) {
      playlists(first: -1, sort: { sortBy: $sortBy, reverse: $reverse }) {
        edges {
          node {
            id
            name
            timeAdded
            stats {
              ...UserStatsFields
            }
          }
        }
      }
    }

    ${UserStatsFields}
  `;

  const { playlists: { edges } } = await client.request(print(query), {
    sortBy,
    reverse,
  });

  return edges.map(({ node }) => node);
});

testSort('playlist items', async ({ sortBy, reverse }) => {
  const query = gql`
    query($playlistId: ID!, $sortBy: SortBy!, $reverse: Boolean!) {
      playlist(id: $playlistId) {
        items(first: -1, sort: { sortBy: $sortBy, reverse: $reverse }) {
          edges {
            node {
              song {
                id
                name
                timeAdded
                stats {
                  ...UserStatsFields
                }
              }
            }
          }
        }
      }
    }

    ${UserStatsFields}
  `;

  const { playlist: { items: { edges } } } = await client.request(
    print(query),
    {
      sortBy,
      reverse,
      playlistId: uuidForNum(3),
    }
  );

  return edges.map(({ node: { song } }) => song);
});
