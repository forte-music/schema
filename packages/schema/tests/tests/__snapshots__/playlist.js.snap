// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get a playlist by id 1`] = `
Object {
  "playlist": Object {
    "description": "",
    "duration": 481,
    "id": "00000000000000000000000000000001",
    "items": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "1:0",
            "song": Object {
              "diskNumber": 1,
              "id": "00000000000000000000000000000002",
              "name": "I'm That (Remix)",
              "streamUrl": "/music/track.mp3",
              "trackNumber": 1,
            },
          },
        },
        Object {
          "node": Object {
            "id": "1:1",
            "song": Object {
              "diskNumber": 1,
              "id": "00000000000000000000000000000003",
              "name": "Bugatti",
              "streamUrl": "/music/track.mp3",
              "trackNumber": 1,
            },
          },
        },
      ],
    },
    "name": "Quick Jam Session",
  },
}
`;
